snapcraft/lifecycle.py

```diff
@@ -48,10 +49,24 @@ def execute(step, part_names=None):
         parts = {p for p in config.all_parts if p.name in part_names}
     else:
         parts = config.all_parts
+        part_names = config.part_names
 
     step_index = common.COMMAND_ORDER.index(step) + 1
     for step in common.COMMAND_ORDER[0:step_index]:
+        if step == 'stage':
+            _check_for_collisions(config.all_parts)
         for part in parts:
+            prereqs = config.part_prereqs(part.name)
+            if prereqs and not prereqs.issubset(part_names):
```

Comment:
what if you ran the prereqs before?
Like on the test you added, you could have:
lifecycle.execute('stage', part_names=['part1'])
lifecycle.execute('pull', part_names=['part2'])

should that succeed?